proposals/DOM-Parts-Declarative.md

```diff
@@ -0,0 +1,128 @@
+# DOM Parts Declarative API
+
+This proposal covers the declarative API for creating a DOM part within a
+`<template>` element and main document HTML.
+
+## Proposal
+
+Double curly braces `{{}}` provide markers for DOM parts.
+
+In the most basic level, this proposal can produce the three DOM parts:
+`NodePart`, `AttributePart`, `ChildNodePart`.
+
+### Basic Examples
+
+Suppose we had the following template in some HTML extension template languages
+where `{name}` and `{email}` indicated locations of dynamic data insertion:
+
+```html
+<section>
+  <h1 id="name">{name}</h1>
+  Email: <a id="link" href="mailto:{email}">{email}</a>
+</section>
+```
+
+And the application has produced an HTML `<template>` with the following
+content:
+
+```html
+<template>
+  <section>
+    <h1 id="name">{{}}</h1>
+    Email: <a id="link" href="{{}}">{{}}</a>
+  </section>
+</template>
+```
+
+This will create a `ChildNodePart` attached to `<h1>` with no content, an
+`AttributePart` connected to `href`, and a `ChildNodePart` connected to `<a>`
+with no content.
+
+A framework could fetch these parts use either a
+[`DocumentPart`](./DOM-Parts-Imperative.md#option-2-documentpartgroup) on the
+`DocumentFragment` produced by the template.
+
+## Enablement in main `document`
+
+For the main document, a new `parseparts` attribute is introduced that indicates
+to the parser it should parse DOM part tags as DOM parts.
+
+```html
+<div parseparts></div>
+```
+
+Even for `innerHTML` use cases, only DOM nodes that are wrapped DOM with the
+`parseparts` attribute will use declarative parts.
+
+## Node parts
+
+For node parts, a `{{}}` tag could be provided as an attribute.
+
+```html
+<template>
+  <section {{}}></section>
+</template>
+```
+
+Would create a `NodePart` for `<section>`.
+
+## Partial attributes
+
+Allowing `{{}}` inside an attribute works the same as a
+[partial attribute update](./DOM-Parts-Imperative.md#partial-attribute-updates),
+in that it will create an `AttributePart` for the entire attribute, but it will
+have multi-valued `value` property.
+
+```html
+<template>
+  <section>
+    <h1 id="name">{{}}</h1>
+    Email: <a id="link" href="mailto:{{}}">{{}}</a>
+  </section>
+</template>
+```
+
+The `AttributePart` for `href` would have `value` equal to `['mailto:', '']`.
```

Comment:
Oh I was imagining we would replace the value field entirely. Because otherwise you could never update what otherwise looked like a static string?

Reply:
Two motivating examples:

```html
  <div class="foo {{color}} bar"></div>
  <iframe src="https://partner.example.com/page?{{params}}">
```

In these cases, I think a user would expect that after updating these AttributePart's values, foo and bar would still be classes on the div, and the iframe would only load from the partner page.

If a library on top of template parts wants to update the attribute to a value that doesn't include the static strings (e.g. via a user level directive system, or when set to a magic value defined by the library), it could still do so by getting the element and attribute name off the part and calling imperative APIs.

Reply:
Okay, makes sense. I added a new `statics` attribute to `AttributePart` with declarative or imperative statics.